lib/matplotlib/patches.py

```diff
@@ -2551,37 +2669,157 @@ class DArrow:
-            # padding
+            # scaled padding
             pad = mutation_size * self.pad
-            # width and height with padding added.
-            # The width is padded by the arrows, so we don't need to pad it.
+            # add padding to height
             height = height + 2 * pad
-            # boundary of the padded box
+            # boundary points of the padded box (arrow tail/body)
             x0, y0 = x0 - pad, y0 - pad
             x1, y1 = x0 + width, y0 + height
-
+            # half-width and quarter-width of arrow tail
             dx = (y1 - y0) / 2
             dxx = dx / 2
-            x0 = x0 + pad / 1.4  # adjust by ~sqrt(2)
-
-            return Path._create_closed([
-                (x0 + dxx, y0), (x1, y0),  # bot-segment
-                (x1, y0 - dxx), (x1 + dx + dxx, y0 + dx),
-                (x1, y1 + dxx),  # right-arrow
-                (x1, y1), (x0 + dxx, y1),  # top-segment
-                (x0 + dxx, y1 + dxx), (x0 - dx, y0 + dx),
-                (x0 + dxx, y0 - dxx),  # left-arrow
-                (x0 + dxx, y0)])
+
+            x0 += pad / 1.4  # adjust by ~sqrt(2)
+
+            # The width adjustment is the distance that must be subtracted from
+            # y0 and added to y1 to reach the non-tip vertices of the head.
+            # The body width is 2dx.
+            # Subtracting 1 from the head width gives, in units of the body width,
+            # the total 'width' of arrow-head not within the body.
+            width_adjustment = (self.head_width - 1) * dx
+
+            if self.head_angle == 0:
+                # This would cause a division by zero ('infinitely long' arrow head)
+                raise ValueError("Head angle of zero is not valid.")
+            elif self.head_angle <= 180:
+                # Non-reversed arrow heads (<--->)
+
+                # tan(1/2 * angle subtended by arrow tip)
+                tan_half_angle = np.tan(self.head_angle * (math.pi / 360))
+
+                # The angle adjustment is the tip-to-body length of the arrow head.
+                # Each half of the arrow head is a right-angled triangle. Therefore,
+                # each half of the arrow head has, by trigonometry, tan(head_angle/2)=
+                # (dx+width_adjustment)/(dxx+angle_adjustment).
+                angle_adjustment = ((dx + width_adjustment) / tan_half_angle) - dxx
+
+                # Check padding; the ends of the text should have a minimum
+                # clearance of pad from the heads
+                if -width_adjustment < pad:
+                    # Only do this if the text fits into the head
+                    text_clearance = (width_adjustment / tan_half_angle) + \
+                        pad * ((1 / tan_half_angle) - 1)
+
+                    if text_clearance < pad:
+                        # Lengthen arrow body to accommodate text
+                        x0 += text_clearance - pad
+                        x1 += pad - text_clearance
```

Comment:
I specifically wrote it the other way for clarity; `x0` is reduced by the same amount that `x1` is increased just as in the other cases.
```suggestion
                        x0 -= pad - text_clearance
                        x1 += pad - text_clearance
```